cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp

```diff
@@ -59,6 +61,171 @@ inline sycl::event beta_copy_transform(sycl::queue& queue,
     });
 }
 
+/*
+This is an adaptation of the CPU version, which can be found in this file:
+cpp/daal/src/algorithms/service_kernel_math.h
+
+It solves a linear system A*x = b
+where 'b' might be a matrix (multiple RHS)
+
+It is intended as a fallback for solving linear regression, where these
+matrices are obtained as follows:
+    A = t(X)*X
+    b = t(X)*y
+
+It can handle systems that are not positive semi-definite, so it's used
+as a fallback when Cholesky fails or when it involves too small numbers
+(which tends to happen when floating point error results in a slightly
+positive matrix when it should have zero determinant in theory).
+*/
+template <mkl::uplo uplo, typename Float>
+sycl::event solve_spectral_decomposition(
+    sycl::queue& queue,
+    ndview<Float, 2>& A, // t(X)*X, LHS, gets overwritten
+    sycl::event& event_A,
+    ndview<Float, 2>& b, // t(X)*y, RHS, solution is outputted here
+    sycl::event& event_b,
+    const std::int64_t dim_A,
+    const std::int64_t nrhs) {
+    constexpr auto alloc = sycl::usm::alloc::device;
+
+    /* Decompose: A = Q * diag(l) * t(Q) */
+    /* Note: for NRHS>1, this will overallocate in order to reuse the memory as buffer later on */
+    auto eigenvalues = ndarray<Float, 1>::empty(queue, dim_A * nrhs, alloc);
+    sycl::event syevd_event =
+        syevd<mkl::job::vec, uplo, Float>(queue, dim_A, A, dim_A, eigenvalues, { event_A });
+    const Float eps = std::numeric_limits<Float>::epsilon();
+
+    /* Discard too small singular values */
+    std::int64_t num_discarded;
+    {
+        /* This is placed inside a block because the array created here is not used any further */
+        /* Note: the array for 'eigenvalues' is allocated with size [dimA, nrhs],
+        but by this point, only 'dimA' elements will have been written to it. */
+        auto eigenvalues_cpu =
+            ndview<Float, 1>::wrap(eigenvalues.get_data(), dim_A).to_host(queue, { syevd_event });
+        const Float* eigenvalues_cpu_ptr = eigenvalues_cpu.get_data();
+        const Float largest_ev = eigenvalues_cpu_ptr[dim_A - 1];
+        if (largest_ev <= eps) {
+            throw std::runtime_error(
+                "Could not solve linear system. Problem has too small singular values.");
```

Comment:
Please add new error message into `error_handling.hpp` and use it like it is done in other algorithms:
https://github.com/oneapi-src/oneDAL/blob/main/cpp/oneapi/dal/algo/knn/backend/gpu/train_kernel_brute_force_dpc.cpp#L38

Reply:
Moved the error message to the file with internal error messages.